Added scripts needed for analysis of WES-data.

monikaBrandt opened the pull request:
- Added scripts used to prepare the analysis with Sarek.
- Added scripts used to get additional metrics after analysis with
Sarek.
- Added templates and python-scripts called/used by bash scripts.
- Added an example script on how to run bedToIntervalList.

monikaBrandt committed:
Added scripts needed for analysis of WES-data.
- Added scripts used to prepare the analysis with Sarek.
- Added scripts used to get additional metrics after analysis with
Sarek.
- Added templates and python-scripts called/used by bash scripts.
- Added an example script on how to run bedToIntervalList.

monikaBrandt committed:
Make scripts executable by changing permissions

alvaannett reviewed (dismissed):
Looks good!

b97pla commented on 1_create_reference_tsv.bash:
```diff
@@ -0,0 +1,32 @@
+
+echo "     Project: ${project}"
+
+if [ ! -d "/proj/ngi2016001/nobackup/NGI/DATA/${project}" ]; then
```
for readability and to save some typing, I would recommend using a variable for the DATA and ANALYSIS paths

b97pla commented on 1_create_reference_tsv.bash:
```diff
@@ -0,0 +1,32 @@
+echo "Will create tsv file for sarek in /proj/ngi2016001/nobackup/NGI/ANALYSIS/${project}/reference_tsv."
+
+mkdir -p /proj/ngi2016001/nobackup/NGI/ANALYSIS/${project}/reference_tsv
+find /proj/ngi2016001/nobackup/NGI/DATA/${project} | grep "_R[1,2]_001" | sort > /proj/ngi2016001/nobackup/NGI/ANALYSIS/${project}/reference_tsv/temp.txt
```
alternatively, to drop the `grep` you can always use a file pattern with find, i.e. `find /proj/ngi2016001/nobackup/NGI/DATA/${project} -name "*_R[1,2]_001*"`

b97pla commented on 2_create_twist_exome_analysis.bash:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+SCRIPTSDIR="/proj/ngi2016001/private/local_scripts/exome_scripts"
```
should this really point to local scripts? if you want to point to the same folder that this script is deployed to, you can work with `$0` (a reference to the current script), e.g. `SCRIPTSDIR="$(dirname "$(readlink -f "$0")")"`

b97pla replied:
Note though that this doesn't work well if the script is executing on a compute node through slurm since in that case, the script is copied by slurm to a local location.

b97pla commented on bed2interval_list.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+module load bioinfo-tools picard
+java -jar /lupus/sw/bioinfo/picard/2.10.3/milou/picard.jar BedToIntervalList \
```
the path to the Picard jar should not be hard coded, use `java -jar ${PICARD_HOME}/picard.jar` instead

b97pla commented on bed2interval_list.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+module load bioinfo-tools picard
+java -jar /lupus/sw/bioinfo/picard/2.10.3/milou/picard.jar BedToIntervalList \
+      I=/proj/ngi2016001/private/local_scripts/exome_scripts/Twist_Exome_RefSeq_targets_hg19.bed \
```
same comment as above regarding local scripts

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,38 @@
+		else:
+			newline = '\t'.join((sample,"ZZ","0",sample,str(sampleDict[sample]),i))
+
+		with open(outFile, 'a') as of:
```
I would keep the opening, writing and closing to the output file outside the loop and write everything in one go at the end

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,38 @@
+k = 0
+for i in lines:
+	sample = i.split('/')[-1].split('_')[0]
+	readnr = i.split('/')[-1].split('_')[3]
```
I think these assignments are pretty vulnerable. If a sample name contains one or more underscores, the logic won't work.
A more reliable approach would be to use a regexp, e.g. something like: 
```suggestion
        import re
	pattern = r'^(.*)_(S\d+)_L(\d+)_(R\d)_001.*$'
	m = re.match(os.path.basename(i), pattern)
	if m:
	  sample = m.group(1)
	  ssheet_idx = m.group(2)
	  laneno = int(m.group(3))
	  readnr = m.group(4)
	else:
	  continue
```

monikaBrandt replied:
@b97pla As it is now underscores are not allowed in sample name because of a "feature" in bcl2fastq. However, it might be worth changing anyway to make the script more secure.

b97pla replied:
ah ok, yes I would go for the regex approach regardless

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,38 @@
+		with open(outFile, 'a') as of:
+        		of.write(newline)
+
+	elif sample not in sampleDict and readnr == "R1":
```
afaict, this elif-statement more or less duplicates the code in the if-statement above. it would be nice if this could be avoided. I think the standard approach is to use `collections.defaultdict` but otherwise I would just do 
```
if readnr == "R1":
  if sample not in sampleDict:
    sampleDict[sample] = 0
  sampleDict[sample] += 1
```
and then get rid of one of the statements

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,38 @@
+		if k != len(lines)-1:
+                	nextread = lines[k+1].split('/')[-1].split('_')[3]
+		if nextread == "R2":
+			newline = '\t'.join((sample,"ZZ","0",sample,str(sampleDict[sample]),i.rstrip("\n"),lines[k+1]))
```
I'm not entirely sure if there is a "right" way, but I think the 5th field will be used as a read tag in e.g. the BAM files. Therefore it could be used to hold some information (in addition to separating the reads). In the WGS tsv, this is a string with `"[FCID].[laneno].[ssheet_idx]"`.

The FCID could be extracted from the fastq file path.

monikaBrandt replied:
I can not find anything about this in the documentation. But if we think that this feature is missing it is fairly easy to add in a new PR.

b97pla replied:
This is a good reference for how GATK uses this information and what it should contain: https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups

If you look at the Sarek code here: https://github.com/nf-core/sarek/blob/2.7/main.nf#L1184, you can see that Sarek assigns the same value (equal to the 5th field above) for `ID` and `PU`, which is why I think it would be good to format this according to above.

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,38 @@
+		with open(outFile, 'a') as of:
+                        of.write(newline)
+	k= k+1
+
```
If you want, I can suggest some more edits to clean up the code a bit. Otherwise, it should be ok with the edits above.

b97pla commented on run_hs_metrics.sh:
```diff
@@ -0,0 +1,66 @@
+METRICS=$5
+
+# output folder for metrics and logs 
+OUTDIR="${ANALYSISDIR}/results/Reports/HsMetrics"
```
just a note that, for 2.7, the logs are under `OUTDIR="${ANALYSISDIR}/results/Reports/[SAMPLE]/HsMetrics"` so this needs to be updated when the pipeline is changed

monikaBrandt replied:
I'm not sure what you mean with this comment @b97pla. The OUTDIR should be independent of the pipeline output structure. Of course we can change the structure of the output if we want to, but I can't see why it is a must.

b97pla replied:
I was thinking that we want to harmonize the output so that the HsMetrics report for a sample will be put alongside the other QC reports from the pipeline for that sample. Not necessarily a must I guess..

b97pla reviewed (changes requested):
great work!

this looks mostly good and my comments are nitpicks, though `create_reference_tsv.py` requires some changes as far as I can tell 😃

monikaBrandt committed:
Change paths to variables

monikaBrandt committed:
Removed unecessary code and introduced re.match

monikaBrandt commented:
@sarek928 and @matrulda I have now addressed the comments I've got from @b97pla. Somethings I have not changed because I find it unnecessary at this time. However, I will keep it in mind for the future. Where I have chosen to not implement requested changes, an explanatory comment should be provided below. Happy reviewing!

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+k = 0
+for i in lines:
+	pattern = r'^(.*)_(S\d+)_L(\d+)_(R\d)_001.*$'
+	bnm = re.match(pattern, os.path.basename(i))
```
It is not clear to me what `bnm` stands for.

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+for i in lines:
+	pattern = r'^(.*)_(S\d+)_L(\d+)_(R\d)_001.*$'
+	bnm = re.match(pattern, os.path.basename(i))
+	#TODO: 5th field in tsv might later on be changed from an int to a string containing "[FCID].[laneno].[ssheet_idx]".
```
I think this TODO would be clearer if it's stated in the beginning that the TODO is about also fetching the flowcell ID.

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+outList = []
+
+with open(inFile, 'r') as f:
+	lines = f.readlines()
```
According to PEP8 it is recommended to use 4 spaces instead tab as indentation level. Personally I also find it more readable that way.

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+
+		if k != len(lines)-1:
+			nxtrd = re.match(pattern, os.path.basename(lines[k+1]))
+		if nxtrd.group(4) == "R2":
```
Speaking as a sarek n00b, I think it would be helpful to add a comment below this (or somewhere else) that the tsv should contain these columns: `subject sex status sample lane fastq1 fastq2`

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+	if readnr == "R1":
+		if sample not in sampleDict:
+			sampleDict[sample] = 0
+		sampleDict[sample] += 1
```
Why isn't lane number used instead?

matrulda commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+		k += 1
+		continue
+
+with open(outFile, 'a') as of:
```
Since you are only writing once I think it would be clearer to use write mode 'w'.

matrulda reviewed (dismissed):
Looks good, left some comments and suggestions.

b97pla commented on create_reference_tsv.py:
```diff
@@ -0,0 +1,48 @@
+	bnm = re.match(pattern, os.path.basename(i))
+	#TODO: 5th field in tsv might later on be changed from an int to a string containing "[FCID].[laneno].[ssheet_idx]".
+	# Only applicable on NovaSeq runs. Split below will not work for Miseq runs.
+	#fcid = i.split(os.path.sep)[-2].split("_")[-1][1:]
```
something like this would hopefully work:

```suggestion
    pattern = r'^.*_[AB]?([^_]+)$'
    m = re.match(pattern, os.path.basename(os.path.dirname(i)))
    fcid = m.group(1) if m else None
```

monikaBrandt committed:
Changed logics and output with additional re

monikaBrandt commented:
New commit available @b97pla and @matrulda! I hope you like it :)

matrulda reviewed (approved):
Nice, good job, looks good to me! 👍

b97pla commented on bed2interval_list.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+module load bioinfo-tools picard
+java -jar ${PICARD_HOME}/picard.jar BedToIntervalList \
```
Noted that this is just an example script but maybe it would be good to clarify this with an inline comment, in particular that the file names refer to hg19 (perhaps mostly this was a bit unexpected since the other scripts assume GRCh38..). I would perhaps suggest also to rename the file to e.g. `bed2interval_list.example.sh`

b97pla reviewed (approved):
Great work :+1:. I think this is good to merge if you're happy. I left some comment that you can fix if you agree. 

Another improvement(?) could be to actually set up the analysis so that CollectHsMetrics is always run after the pipeline, thus eliminating a manual step. This could be done either directly from the analysis sbatch file or from another script that submits both the analysis and then the HsMetrics jobs with dependencies on the analysis.

monikaBrandt commented:
Thank you @b97pla and @matrulda for you reviews and valuable input. I will leave this PR as it is right now, but I will have your comments in mind next time it is time for an update. Can someone please merge this?
